node/setup.go

```diff
@@ -528,67 +528,60 @@ func startStateSync(
 
 //------------------------------------------------------------------------------
 
-var genesisDocKey = []byte("genesisDoc")
+var genesisDocHashKey = []byte("genesisDocHash")
 
-// LoadStateFromDBOrGenesisDocProvider attempts to load the state from the
+// loadStateFromDbOrGenesisDoc attempts to load the state from the
 // database, or creates one using the given genesisDocProvider. On success this also
 // returns the genesis doc loaded through the given provider.
-func LoadStateFromDBOrGenesisDocProvider(
+func loadStateFromDbOrGenesisDoc(
+	stateStore sm.Store,
 	stateDB dbm.DB,
-	genesisDocProvider GenesisDocProvider,
-) (sm.State, *types.GenesisDoc, error) {
-	// Get genesis doc
-	genDoc, err := loadGenesisDoc(stateDB)
-	if err != nil {
-		genDoc, err = genesisDocProvider()
-		if err != nil {
-			return sm.State{}, nil, err
-		}
-
-		err = genDoc.ValidateAndComplete()
-		if err != nil {
-			return sm.State{}, nil, fmt.Errorf("error in genesis doc: %w", err)
-		}
-		// save genesis doc to prevent a certain class of user errors (e.g. when it
-		// was changed, accidentally or not). Also good for audit trail.
-		if err := saveGenesisDoc(stateDB, genDoc); err != nil {
-			return sm.State{}, nil, err
-		}
-	}
-	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
-		DiscardABCIResponses: false,
-	})
-	state, err := stateStore.LoadFromDBOrGenesisDoc(genDoc)
+	genDoc *types.GenesisDoc,
+) (sm.State, error) {
+	// 1. Verify genesisDoc hash in db if exists
+	genDocHash, err := stateDB.Get(genesisDocHashKey)
 	if err != nil {
-		return sm.State{}, nil, err
+		return sm.State{}, err
 	}
-	return state, genDoc, nil
-}
 
-// panics if failed to unmarshal bytes
-func loadGenesisDoc(db dbm.DB) (*types.GenesisDoc, error) {
-	b, err := db.Get(genesisDocKey)
+	genDocBytes, err := cmtjson.Marshal(genDoc)
```

Comment:
I wonder if we can plug in some streaming output backend to `Marshal` that would only compute the hash from the data, instead of populating a potentially huge buffer.

JSON is not a canonical data representation, so any semantically insignificant variation, e.g. added whitespace, in the genesis file would result in a hash mismatch, if not for this approach of hashing the sanitized marshalling output. Is this intentional? In any case it adds some robustness. Maybe it's worth explaining somewhere that the hash we store in the database is not actually the hash of the file content.

Reply:
Agree that hashing based on JSON serialization is brittle. Maybe it would be better to use what we use for (signed) votes and proposals (but without signing). Please find an example [here](https://github.com/cometbft/cometbft/blob/8e6e6ca0b/test/e2e/app/app.go#L664).

This won't totally remove json from the equation, as we still have field `AppState`, which contains the app's state in serialized json format. But that one has been like that for years (battle-tested). Although, if we want to improve on that as well, the moment is now (since we'll need an upgrade strategy anyway).

Reply:
I think the concern with the `AppState` is not great if they use the `cometbft` `genesis` type which they do. So in the SDK they use the `genesis.saveAs` from cometbft, as long as that doesn't change I'd think should be OK.

For example, another thing to keep in mind is that `saveAs` function uses a cometbft `MarshalIdent` function so the exported json is in a format that is very specific. If there's any "outside" process that formats this file (e.g. remove indentation) then every might break
https://github.com/cometbft/cometbft/blob/8e6e6ca0bcc732747e10020be2c5e9f3779e6b0c/types/genesis.go#L50